### ISSUE TITLE
Wrong ephemeris for asteroids after MPCORB database change

### ISSUE THREAD

**Issue by ycao-cit** [brandon-rhodes/pyephem#89]
Recently, the MPCORB database changed its standard epoch. Then pyephem failed to calculate the position of asteroids.

Take Asteroid 2001 BP25 on 2015-11-23.55 UT as an example. With the old and new standard epochs, I calculated its position as follows:

```
import ephem
import datetime

tobs = ephem.Date( datetime.datetime(2015, 11, 23, 13, 12, 13) ) # observation time

obj = ephem.readdb( "2001 BP25 0,e,3.47304,134.76688,294.08950,2.4271529,0.26065039,0.2231376,330.92520,06/27.0/2015,2000,H 16.0,0.15" ) # old orbital elements
obj.compute( tobs )
print "With old orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )

obj = ephem.readdb( "2001 BP25 0,e,3.47303,134.76696,294.10956,2.4270882,0.26066081,0.2231111,23.04368,01/12.0/2016,2000,H 16.1,0.15" ) # new orbital elements
obj.compute( tobs )
print "With new orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
```

Then I obtained the following output:

```
With old orbital elements, ra = 1.879871 rad, dec = 0.30927 rad
With new orbital elements, ra = 1.891992 rad, dec = 0.30874 rad
```

The online IAU MP checker confirmed the coordinates calculated with the old orbital elements.

Therefore, I am wondering: is there some constant in pyephem that should be changed for the new standard epoch?

**Comment by albertw** [comment]
Hi,

I think you may have decoded the epoch in the MPCORB incorrectly which accounts for the error. In the current MPCORB the encoded epoch for that object is K161D. If you alter the epoch in the readdb line to that (01/13.0/2016) you get a closer match to your example with older elements.

Cheers,
~Al

```
import ephem
import datetime

tobs = ephem.Date( datetime.datetime(2015, 11, 23, 13, 12, 13) ) # observation time

obj = ephem.readdb( "2001 BP25 0,e,3.47304,134.76688,294.08950,2.4271529,0.26065039,0.2231376,330.92520,06/27.0/2015,2000,H 16.0,0.15" ) # old orbital elements
obj.compute( tobs )
print "With old orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )

obj = ephem.readdb( "2001 BP25 0,e,3.47303,134.76696,294.10956,2.4270882,0.26066081,0.2231111,23.04368,01/12.0/2016,2000,H 16.1,0.15" ) # new orbital elements
obj.compute( tobs )
print "With new orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
                  
obj = ephem.readdb( "2001 BP25 0,e,3.47303,134.76696,294.10956,2.4270882,0.26066081,0.2231111,23.04368,01/13.0/2016,2000,H 16.1,0.15" ) # new orbital elements
obj.compute( tobs )
print "With new & corrected epoc, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
```
```
 With old orbital elements, ra = 1.879871 rad, dec = 0.30927 rad 
 With new orbital elements, ra = 1.891992 rad, dec = 0.30874 rad
 With new & corrected epoc, ra = 1.879874 rad, dec = 0.30927 rad
```

**Comment by ycao-cit** [comment]
Hi,

Thanks a lot for pointing out this bug! I used a code from somebody else to
convert the MPCORB to the Xephem format.

Now it works correctly after correcting this bug.

Best,
Yi

On Wed, Nov 25, 2015 at 3:22 PM, Albert White <notifications@github.com>
wrote:

> Hi,
>
> I think you may have decoded the epoch in the MPCORB incorrectly which
> accounts for the error. In the current MPCORB the encoded epoch for that
> object is K161D. If you alter the epoch in the readdb line to that
> (01/13.0/2016) you get a closer match to your example with older elements.
>
> Cheers,
> ~Al
>
> import ephem
> import datetime
>
> tobs = ephem.Date( datetime.datetime(2015, 11, 23, 13, 12, 13) ) # observation time
>
> obj = ephem.readdb( "2001 BP25 0,e,3.47304,134.76688,294.08950,2.4271529,0.26065039,0.2231376,330.92520,06/27.0/2015,2000,H 16.0,0.15" ) # old orbital elements
> obj.compute( tobs )
> print "With old orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
>
> obj = ephem.readdb( "2001 BP25 0,e,3.47303,134.76696,294.10956,2.4270882,0.26066081,0.2231111,23.04368,01/12.0/2016,2000,H 16.1,0.15" ) # new orbital elements
> obj.compute( tobs )
> print "With new orbital elements, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
>
> obj = ephem.readdb( "2001 BP25 0,e,3.47303,134.76696,294.10956,2.4270882,0.26066081,0.2231111,23.04368,01/13.0/2016,2000,H 16.1,0.15" ) # new orbital elements
> obj.compute( tobs )
> print "With new & corrected epoc, ra = %.6f rad, dec = %.5f rad" % ( obj.a_ra, obj.a_dec )
>
>  With old orbital elements, ra = 1.879871 rad, dec = 0.30927 rad
>  With new orbital elements, ra = 1.891992 rad, dec = 0.30874 rad
>  With new & corrected epoc, ra = 1.879874 rad, dec = 0.30927 rad
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brandon-rhodes/pyephem/issues/89#issuecomment-159753744>
> .
>



-- 
Yi Cao
Astronomy Department, Caltech,
Pasadena, CA 91125
http://www.astro.caltech.edu/~ycao

**Comment by brandon-rhodes** [comment]
Thank you so much, @albertw, for your help on this issue! And please let us know, @ycao-cit, if you have any further problems with PyEphem.
